ovf/ova-compose.py

```diff
@@ -0,0 +1,746 @@
+        item.append(self.xml_element('InstanceID', self.instance_id))
+        item.append(self.xml_element('Description', self.description))
+        item.append(self.xml_element('ElementName', element_name))
+
+        return item
+
+
+class RasdCpus(RasdItem):
+    resource_type = 3
+    description = "Virtual CPUs"
+
+
+    def __init__(self, num):
+        super().__init__()
+        self.num = num
+
+    @classmethod
+    def from_dict(cls, d):
+        item = cls(d)
+        return item
+
+
+    def xml_item(self, required, element_name):
+        item = super().xml_item(required, element_name)
+        item.append(self.xml_element('AllocationUnits', 'hertz * 10^6'))
+        item.append(self.xml_element('VirtualQuantity', self.num))
+        
+        return item
+
+
+class RasdMemory(RasdItem):
+    resource_type = 4
+    description = "Virtual Memory"
+
+
+    def __init__(self, size):
+        super().__init__()
+        self.size = size
+
+
+    @classmethod
+    def from_dict(cls, d):
+        item = cls(d)
+        return item
+
+
+    def xml_item(self, required, element_name):
+        item = super().xml_item(required, element_name)
+        item.append(self.xml_element('AllocationUnits', 'byte * 2^20'))
+        item.append(self.xml_element('VirtualQuantity', self.size))
+        
+        return item
+
+
+class RasdController(RasdItem):
+
+    def __init__(self, subtype):
+        super().__init__()
+        self.next_child_address = 0
+        self.subtype = subtype
+
+
+    @classmethod
+    def from_dict(cls, d):
+        item = cls(d.get('subtype', None))
+        subtype = d.get('subtype', None)
+        return item
+
+
+    def add_child(self, child):
+        child.address_on_parent = self.next_child_address
+        self.next_child_address += 1
+
+
+class RasdScsiController(RasdController):
+    resource_type = 6
+    description = "SCSI Controller"
+
+
+    def __init__(self, subtype):
+        super().__init__(subtype)
+        self.next_child_address = 0
+        # TODO: maintain valid settings in a structure instead of code:
+        if self.subtype is None:
+            self.subtype = 'VirtualSCSI'
+        elif self.subtype.lower() in ['virtualscsi', 'lsilogic']:
+            self.subtype = subtype
+        else:
+            raise Exception(f"invalid SCSI subtype '{self.subtype}'")
+
+
+    def xml_item(self, required, element_name):
+        item = super().xml_item(required, element_name)
+        item.append(self.xml_element('ResourceSubType', self.subtype))
+        
+        return item
+
+
+class RasdSataController(RasdController):
+    resource_type = 20
+    description = "SATA Controller"
+
+    def xml_item(self, required, element_name):
+        item = super().xml_item(required, element_name)
+        item.append(self.xml_element('ResourceSubType', 'vmware.sata.ahci'))
+        
+        return item
+
+
+class RasdIdeController(RasdController):
+    resource_type = 4
+    description = "IDE Controller"
```

Comment:
After resolving network section, it reported another error:
```
Issues detected with selected template. Details: - 40:7:ELEMENT_ILLEGAL: Multiple ''memory'' elements are not allowed. - 77:7:VALUE_ILLEGAL: Value ''5'' of Parent element does not refer to a ref of type DiskControllerReference.
```

```
      <Item>
        <rasd:Description>IDE Controller</rasd:Description>
        <rasd:ElementName>ide1</rasd:ElementName>
        <rasd:InstanceID>5</rasd:InstanceID>
        <rasd:ResourceType>4</rasd:ResourceType>
      </Item>
```
By comparing ovf template, the resource type of IDE controller shall be 5.